src/stenographer.c

```diff
@@ -0,0 +1,372 @@
+        ud.data = st;
+        ud.free_func = (void *) stenographer_free;
+
+        memset (callback_name, 0, sizeof (callback_name));
+        ssnprintf (callback_name, sizeof (callback_name),
+                "stenographer/%s/%s",
+                (st->host != NULL) ? st->host : hostname_g,
+                (st->name != NULL) ? st->name : "default"),
+
+        status = plugin_register_complex_read (/* group = */ NULL,
+                /* name      = */ callback_name,
+                /* callback  = */ stenographer_read_host,
+                /* interval  = */ NULL,
+                /* user_data = */ &ud);
+    }
+
+    if (status != 0)
+    {
+        stenographer_free(st);
+        return (-1);
+    }
+
+    return (0);
+} /* int config_add */
+
+static int config (oconfig_item_t *ci)
+{
+    int status = 0;
+    int i;
+
+    for (i = 0; i < ci->children_num; i++)
+    {
+        oconfig_item_t *child = ci->children + i;
+
+        if (strcasecmp ("Instance", child->key) == 0)
+            config_add (child);
+        else
+            WARNING ("stenographer plugin: The configuration option "
+                    "\"%s\" is not allowed here. Did you "
+                    "forget to add an <Instance /> block "
+                    "around the configuration?",
+                    child->key);
+    } /* for (ci->children) */
+
+    return (status);
+} /* int config */
+
+/* initialize curl for each host */
+static int init_host (stenographer_t *st) /* {{{ */
+{
+    assert (st->url != NULL);
+    /* (Assured by `config_add') */
+
+    if (st->curl != NULL)
+    {
+        curl_easy_cleanup (st->curl);
+        st->curl = NULL;
+    }
+
+    if ((st->curl = curl_easy_init ()) == NULL)
+    {
+        ERROR ("stenographer plugin: init_host: `curl_easy_init' failed.");
+        return (-1);
+    }
+
+    curl_easy_setopt (st->curl, CURLOPT_ERRORBUFFER, st->stenographer_curl_error);
+    curl_easy_setopt (st->curl, CURLOPT_NOPROGRESS, 1L);
+    curl_easy_setopt (st->curl, CURLOPT_WRITEFUNCTION, stenographer_curl_callback);
+    curl_easy_setopt (st->curl, CURLOPT_WRITEDATA, st);
+    curl_easy_setopt (st->curl, CURLOPT_URL, st->url);
+    curl_easy_setopt (st->curl, CURLOPT_FOLLOWLOCATION, 1L);
+    curl_easy_setopt (st->curl, CURLOPT_MAXREDIRS, 50L);
+    curl_easy_setopt (st->curl, CURLOPT_SSLCERT, st->cert);
+    curl_easy_setopt (st->curl, CURLOPT_SSLKEY, st->key);
+    curl_easy_setopt (st->curl, CURLOPT_CAINFO, st->cacert);
+    curl_easy_setopt (st->curl, CURLOPT_TCP_KEEPALIVE, 1L);
+
+    return (0);
+} /* }}} int init_host */
+
+static void submit_value (const char *type, const char *type_instance,
+        value_t value, stenographer_t *st)
+{
+    value_list_t vl = VALUE_LIST_INIT;
+
+    vl.values = &value;
+    vl.values_len = 1;
+
+    sstrncpy (vl.host, (st->host != NULL) ? st->host : hostname_g,
+            sizeof (vl.host));
+
+    sstrncpy (vl.plugin, "stenographer", sizeof (vl.plugin));
+    if (st->name != NULL)
```

Comment:
This is guaranteed by the config. Please change to an `assert()` if you want to demonstrate that this can't be NULL.